### ISSUE TITLE
Exploring low-rank matrix factorization (or linear bottleneck) in final affine layers of a neural network

### ISSUE THREAD

**Issue by vijayaditya** [kaldi-asr/kaldi#957]
With the introduction of significantly large LVCSR tasks like fisher+swbd, ASpIRE and multi_en in Kaldi we have been able to train large models (~27 million parameters). Even with the inclusion of chain recipes, where the model is evaluated at 30 Hz frame rate, it is still desirable to reduce the size of model to gain further speed-ups.

Low-rank matrix factorization of final layers (see [1]) is an attractive option for easy speed-ups. The final affine layer in a neural network typically has 40-50% of all the parameters. Reducing the size of the final layer using a linear bottle-neck is hence very desirable (we could also test a non-linear bottle-neck). This can provide significant speed-ups especially during decoding.
e.g. Using a 125 dimensional bottle-neck in the 27 million parameter ASpiRE TDNN model reduces its size to 17 million parameters. 

These experiments just require a 2 or 3 line modifcation in the `steps/nnet3/tdnn/make_configs.py` script. Further, we might not even need to run a lot of tuning experiments, as most comparisons have already been made in [1]. 

[1] T. N. Sainath, B. Kingsbury, V. Sindhwani, E. Arisoy and B. Ramabhadran, "Low-Rank Matrix Factorization for Deep Neural Network Training with High-Dimensional Output Targets," in Proc. ICASSP, May 2013.

Requirements : basic python, understanding of nnet3 recipes

**Comment by vijayaditya** [comment]
We just found that  the matrix multiplies corresponding to the final-affine components in chain models (ie. chain and xent branches) constitute almost 70% of the training time for our typical TDNN models (512x6000 final matrices). So I would like to prioritize this project. 

```
c86: Final-xent_affine.Backprop : time 2.8047679 : fraction 0.233143034181
c76: Final_affine.Backprop : time 2.784232 : fraction 0.231436011637
c73: Final_affine.Propagate : time 1.4015891 : fraction 0.116505446118
c64: Final-xent_affine.Propagate : time 1.396891 : fraction 0.116114922079
```

@GaofengCheng Can you help with this ?

**Comment by GaofengCheng** [comment]
OK,  I will explore it after October 4 ( I'm on vacation for the two days ). is the time ok ?

**Comment by vijayaditya** [comment]
Sure, that should be fine.

**Comment by GaofengCheng** [comment]
@vijayaditya  will the newest version of kaldi show the training time by default or should I do some scripts changes?

**Comment by vijayaditya** [comment]
You can get the info when you run training with --computation.debug=true.
You can summarize the training info by using the script
steps/nnet3/report/summarize*.py.

Vijay

**Comment by GaofengCheng** [comment]
@vijayaditya  I have tested the tdnn_7e_sp (chain) and the tdnn_b_sp (CE) in swbd already, but the time results is far different from yours.  

-  configs: 7e(chain) is 625\*6074 for the final_affine and b(CE) is 1024\*8797 for the final affine.

- time calculate command:
 `grep DebugAfterExecute exp/chain/tdnn_7e_sp/log/train.200.5.log | steps/nnet3/report/summarize_compute_debug_timing.py > timing/tdnn_7e_sp_chain_time.200.5.txt`;
`grep DebugAfterExecute exp/nnet3/tdnn_b_sp/log/train.123.5.log | steps/nnet3/report/summarize_compute_debug_timing.py > timing/tdnn_b_sp_ce_time.123.5.txt`

- time consuming results: 
tdnn_7e_sp(chain): 
`Tdnn_1_affine.Backprop : time 6.0880497 : fraction 0.122792924026 `
`Final_affine.Backprop : time 4.5778669 : fraction 0.0923332906519 `
`Final-xent_affine.Backprop : time 4.5149875 : fraction 0.0910650445358 `
`Tdnn_2_affine.Backprop : time 2.09363414 : fraction 0.0422275557132 `
`Final-xent_affine.Propagate : time 2.04446498 : fraction 0.0412358382954` 
`Tdnn_3_affine.Backprop : time 2.01516592 : fraction 0.0406448908777 `
`Final_affine.Propagate : time 1.99482607 : fraction 0.0402346462544 `
tdnn_b_sp(CE):
`Tdnn_1_affine.Backprop : time 11.3037121 : fraction 0.081469918255`
`Tdnn_2_affine.Backprop : time 11.1117179 : fraction 0.0800861470087`
`Final_affine.Backprop : time 9.4060107 : fraction 0.0677925017954`
`Tdnn_5_affine.Backprop : time 6.67906615 : fraction 0.0481384317334`
`Tdnn_3_affine.Backprop : time 6.50049233 : fraction 0.0468513859922`
`Tdnn_4_affine.Backprop : time 6.0367985 : fraction 0.0435093778014`
`Tdnn_0_affine.Backprop : time 5.47894241 : fraction 0.0394887083391`
`Tdnn_6_affine.Backprop : time 4.17659184 : fraction 0.0301021994172`
`Final_affine.Propagate : time 3.33821903 : fraction 0.0240597450718`

Is my time calculate command wrong leading to the difference or the baseline TDNN scripts causing the difference? If the latter one , could you tell me which TDNN scripts you run in your experiments?

**Comment by danpovey** [comment]
What type of GPU are you using?

**Comment by GaofengCheng** [comment]
@danpovey Tesla K80

**Comment by vijayaditya** [comment]
IIRC I was checking these for tdnn_7b in swbd.

**Comment by GaofengCheng** [comment]
I will test the 7b next

**Comment by vijayaditya** [comment]
@GaofengCheng With the recent xconfig style of network specification these experiments can be checked easily, as they just require changing one line in a local/chain/run_tdnn_*.sh script. 

Could you right now add a run_tdnn_7i.sh script, in swbd recipe, which incorporates the details of your best setup into tdnn_7h. You can check-in the results later.

Please specify the speed gains in the script.

**Comment by GaofengCheng** [comment]
@vijayaditya ok, I will pr a new run_tdnn_7i.sh script today and later update the results.

**Comment by vijayaditya** [comment]
being addressed in #1210
